source/reporters/ut_documentation_reporter.tpb

```diff
@@ -0,0 +1,130 @@
+create or replace type body ut_documentation_reporter is
+
+  constructor function ut_documentation_reporter(self in out nocopy ut_documentation_reporter, a_output ut_output default ut_output_dbms_output()) return self as result is
+  begin
+    self.name               := $$plsql_unit;
+    self.output             := a_output;
+    self.lvl                := 0;
+    self.failed_tests       := ut_objects_list();
+    self.test_count         := 0;
+    self.igonred_test_count := 0;
+    return;
+  end;
+
+  member function tab(self in ut_documentation_reporter) return varchar2 is
+  begin
+    return rpad(' ', self.lvl * 2);
+  end tab;
+
+  overriding member procedure print(self in out nocopy ut_documentation_reporter, a_text varchar2) is
+  begin
+    (self as ut_reporter).print(tab || a_text);
+  end print;
+
+  overriding member procedure before_suite(self in out nocopy ut_documentation_reporter, a_suite ut_object) as
+    l_suite ut_test_suite := treat(a_suite as ut_test_suite);
+  begin
+    self.print( coalesce( a_suite.name, l_suite.object_name ) );
+    lvl := lvl + 1;
+  end;
+
+  overriding member procedure before_test(self in out nocopy ut_documentation_reporter, a_test ut_object) as
+  begin
+    test_count := test_count + 1;
+    if treat(a_test as ut_test).get_ignore_flag() then
+      igonred_test_count := igonred_test_count + 1;
+    end if;
+  end;
+  overriding member procedure after_test(self in out nocopy ut_documentation_reporter, a_test ut_object) as
+    l_test    ut_test := treat(a_test as ut_test);
+    l_message varchar2(4000);
+  begin
+    l_message := coalesce( l_test.name, l_test.test.form_name );
+    --if test failed, then add it to the failures list, print failure with number
+    if l_test.result != ut_utils.tr_success then
+      failed_tests.extend;
+      failed_tests(failed_tests.last) := l_test;
+      l_message := l_message || ' (FAILED - '||failed_tests.last||')';
+    end if;
+    self.print( l_message );
+  end;
+
+  overriding member procedure after_suite(self in out nocopy ut_documentation_reporter, a_suite ut_object) as
+    l_suite ut_test_suite := treat(a_suite as ut_test_suite);
+  begin
+    lvl := lvl - 1;
+    self.print( ' ' );
+  end;
+
+  overriding member procedure after_run(self in out nocopy ut_documentation_reporter, a_suites in ut_objects_list) as
+    l_start_time    timestamp with time zone := to_date('9999','yyyy');
+    l_end_time      timestamp with time zone := to_date('0001','yyyy');
+    procedure print_failures_summary is
```

Comment:
I agree, that is not a best place for this code, as it can be reused.
On the other hand, we can always create reporters on top of this reporter.
Maybe a package `ut_documentation_reporter_helper`?
I can't think of a right place to put this code in.
Anyway, for now I've refactored the procedure and split it into several pieces to separate smaller functionalities.